lib/spack/spack/compilers/msvc.py

```diff
@@ -169,10 +169,12 @@ def setup_custom_environment(self, pkg, env):
             if key and value
         )
 
-        if "path" in int_env:
-            env.set_path("PATH", int_env["path"].split(";"))
-        env.set_path("INCLUDE", int_env.get("include", "").split(";"))
-        env.set_path("LIB", int_env.get("lib", "").split(";"))
+        for env_var in int_env:
+            if int_env[env_var] != os.environ.get(env_var, None):
```

Comment:
> Also a question: should this be appending/prepending vs. setting? I can say that this precedes nearly all modifications to env variables (other than ones of the form `SPACK_...`) with the exception of `CCFLAGS` etc., so it doesn't seem like a major problem at the moment, but depending on where that is moved in the process, it could cause more trouble in the future.

If this changes in the future we will need to change how/when we invoke VCVARS. The intent behind vcvars is that it is the base build env for a given MSVC toolchain, and all other changes should be applied on top of it and take vcvars into consideration. With that in mind, I think set operations are appropriate as opposed to appending/prepending. Additionally prepending/appending only really makes sense for PATH and other list env variables, which, IIRC, vcvars does anyway for variables where that consideration is important.

Reply:
> The intent behind vcvars is that it is the base build env for a given MSVC toolchain, and all other changes should be applied on top of it

Sounds good

> Additionally prepending/appending only really makes sense for PATH and other list env variables, which, IIRC, vcvars does anyway for variables where that consideration is important.

It doesn't look like we propagate such actions (i.e. everything here is `set`): are you saying we ignore all prepend-path actions? Or is it that all such actions are reinterpreted as `set`. Either is fine (in particular given the first point) but I'm checking if I misunderstood something.

Reply:
Sorry, I wasn't clear. We should be doing `set` operations as vcvars sets up the environment correctly itself (including any distinction between prepending, appending, and setting), and our contributions should be to very simply honor those changes and propagate them to Spack.  For example PATH, after running vcvars, will have all of the same entries and the entries added by vcvars, so Spack should just pipe this new path through as vcvars already considered set vs append. FWIW I think this ONLY applies to the path and everything else is a set (maybe LIBPATH too, although that's not set in my base env)